flow/aggregates.py

```diff
@@ -0,0 +1,505 @@
+        other_select = self._get_unique_function_id(other._select)
+        # Get unique id for _aggregator_function attribute
+        self_aggregator_function = self._get_unique_function_id(self._aggregator_function)
+        other_aggregator_function = self._get_unique_function_id(other._aggregator_function)
+
+        return self_select == other_select and \
+            self_aggregator_function == other_aggregator_function
+
+    def __hash__(self):
+        blob_l = [
+            hash(self._sort_by), hash(self._reverse_order),
+            self._get_unique_function_id(self._aggregator_function),
+            self._get_unique_function_id(self._select)
+        ]
+        blob = ','.join(str(attr) for attr in blob_l)
+
+        return int(sha1(blob.encode('utf-8')).hexdigest(), 16)
+
+    def _get_unique_function_id(self, func):
+        """Generate unique id for the function passed. The id returned is used to generate
+        hash and compare arbitrary types which are callable like ``self._aggregator_function``
+        and ``self._select`` attributes.
+        """
+        try:
+            return func.__code__.co_code
+        except Exception:  # Got something other than a function
+            return hash(func)
+
+    def _create_AggregatesStore(self, project):
+        """Create the instance of _AggregatesStore or _DefaultAggregateStore classes
+        to make sure that we don't explicitly store the aggregates in the FlowProject.
+
+        When iterated over these classes, an aggregate is yielded.
+
+        :param project:
+            A signac project used to fetch jobs for creating aggregates.
+        :type project:
+            :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+        """
+        if not self._is_aggregate:
+            return _DefaultAggregateStore(project)
+        else:
+            return _AggregatesStore(self, project)
+
+    def __call__(self, func=None):
+        if callable(func):
+            setattr(func, '_flow_aggregate', self)
+            return func
+        else:
+            raise TypeError('Invalid argument passed while calling '
+                            'the aggregate instance. Expected a callable, '
+                            f'got {type(func)}.')
+
+
+class _AggregatesStore:
+    """This class holds the information of all the aggregates associated with
+    a :py:class:`aggregator`.
+
+    This is a callable class which, when called, generates all the aggregates.
+    When iterated through it's instance, all the aggregates are yielded.
+
+    :param aggregator:
+        aggregator object associated with this class.
+    :type aggregator:
+        :py:class:`aggregator`
+    :param project:
+        A signac project used to fetch jobs for creating aggregates.
+    :type project:
+        :py:class:`flow.FlowProject` or :py:class:`signac.contrib.project.Project`
+    """
+    def __init__(self, aggregator, project):
+        self._aggregator = aggregator
```

Comment:
We don't need to store the aggregator.

Reply:
We compare the `aggregator` hence I think that we should store the aggregator

Reply:
We just need to compare aggregation_ids or hash to determine uniqueness for standard aggregate store object the default is easier with just comparing and using the project directly.